webhook.go

```diff
@@ -529,41 +539,52 @@ func (whsvr *WebhookServer) mutate(ar *v1beta1.AdmissionReview) *v1beta1.Admissi
 	pod.Namespace = podNamespace
 	pod.ObjectMeta.Namespace = podNamespace
 
-	err := isSafe(&pod, string(req.Operation))
+	if isProtectedNamespace(whsvr.protectedNamespaces, &pod.ObjectMeta) {
+
+		// mutation is not allowed in the protected namespace
+		if mutationRequired(&pod.ObjectMeta) {
+			err := ErrProtectedNs
+			glog.Infof("Not safe to continue with %v for pod: %v/%v. Disallowing", req.Operation, pod.GenerateName, pod.Namespace)
+			reason := metav1.StatusReason(Msg1)
+			return &v1beta1.AdmissionResponse{
+				Allowed: false,
+				Result: &metav1.Status{
+					Message: err.Error(),
+					Reason:  reason,
+				},
+			}
+		}
 
-	if err != nil {
-		glog.Error(err.Error())
-		glog.Infof("Not safe to continue with %v for pod: %v/%v. Disallowing", req.Operation, pod.GenerateName, pod.Namespace)
-		reason := metav1.StatusReason("TSI Mutation Webhook disallowed this pod creation for safety reasons")
+		glog.Infof("No mutation requested for %v in %v namespace", pod.GenerateName, pod.Namespace)
 		return &v1beta1.AdmissionResponse{
-			Allowed: false,
-			Result: &metav1.Status{
-				Message: err.Error(),
-				Reason:  reason,
-			},
+			Allowed: true,
 		}
-	}
-	glog.Infof("Safe to continue with %v/%v", pod.GenerateName, pod.Namespace)
 
-	// determine whether to perform mutation
-	required, err := mutationRequired(protectedNamespaces, &pod.ObjectMeta)
-	if err != nil {
-		glog.Error(err.Error())
-		glog.Infof("Not safe to continue with %v for pod: %v/%v. Disallowing", req.Operation, pod.GenerateName, pod.Namespace)
-		reason := metav1.StatusReason("TSI Mutation Webhook disallowed this pod creation for safety reasons")
-		return &v1beta1.AdmissionResponse{
-			Allowed: false,
-			Result: &metav1.Status{
-				Message: err.Error(),
-				Reason:  reason,
-			},
+		// if the namespace is not protected
+	} else {
+
+		// check if the request is safe
+		err := isSafe(&pod, string(req.Operation))
+		if err != nil {
+			glog.Error(err.Error())
+			glog.Infof("Not safe to continue with %v for pod: %v/%v. Disallowing", req.Operation, pod.GenerateName, pod.Namespace)
+			reason := metav1.StatusReason(Msg1)
+			return &v1beta1.AdmissionResponse{
+				Allowed: false,
+				Result: &metav1.Status{
+					Message: err.Error(),
+					Reason:  reason,
+				},
+			}
 		}
```

Comment:
I think it would be nice to have the entire admission logic on whether the request is okay be a single function call? This function is getting a bit big